File: lint-arwaky/src-python/capabilities/semantic_scope_analyzer.py
"""semantic_scope_analyzer — AST-based semantic scope analysis capability."""

from __future__ import annotations
from ..taxonomy import (
    CallChainList,
    Count,
    DataFlowList,
    DirectoryPath,
    FileContentVO,
    FilePath,
    LineNumber,
    ResponseData,
    ResponseDataList,
    ScopeRef,
    SymbolName,
    SymbolNameList,
)

import ast
import re

from ..contract import ISemanticTracerProtocol
from ..contract import IFileSystemPort


class SemanticScopeAnalyzer(ISemanticTracerProtocol):
    """AST-based semantic scope analyzer for Python code."""

    def __init__(self, fs: IFileSystemPort):
        self._fs = fs

    def get_variant_dict(self, name: SymbolName) -> ResponseData:
        n = str(name)
        words = re.findall(r"[A-Za-z][a-z0-9]*|[A-Z]+(?=[A-Z][a-z0-9]|\b)|[0-9]+", n)
        words = [w.lower() for w in words]
        if not words:
            return ResponseData(
                value={
                    "snake_case": n,
                    "pascal_case": n,
                    "camel_case": n,
                    "screaming_snake": n.upper(),
                }
            )
        snake_case = "_".join(words)
        _rest = "".join(str(w).capitalize() for w in words[1:])
        return ResponseData(
            value={
                "snake_case": snake_case,
                "camel_case": str(words[0]) + _rest,
                "pascal_case": "".join(str(w).capitalize() for w in words),
                "screaming_snake": snake_case.upper(),
            }
        )

    def build_variants(self, name: SymbolName) -> SymbolNameList:
        n = str(name)
        d = self.get_variant_dict(name)
        kebab = d.value["snake_case"].replace("_", "-")
        variants = list(
            {
                n,
                d.value["snake_case"],
                d.value["camel_case"],
                d.value["pascal_case"],
                d.value["screaming_snake"],
                kebab,
            }
        )
        return SymbolNameList(values=[SymbolName(value=v) for v in variants])

    def get_enclosing_scope(
        self, file_path: FilePath, line: LineNumber
    ) -> ScopeRef | None:
        if not self._fs.exists(file_path).value:
            return None

        source_vo: FileContentVO = self._fs.read_text(file_path)
        source = source_vo.value

        try:
            tree = ast.parse(source)
        except SyntaxError:
            return None

        class ScopeVisitor(ast.NodeVisitor):
            def __init__(self, target_line: LineNumber):
                self.target_line = int(target_line)
                self.current_path: list[str] = []
                self.best_match: list[str] = []

            def generic_visit(self, node):
                is_scope = isinstance(
                    node, (ast.FunctionDef, ast.AsyncFunctionDef, ast.ClassDef)
                )
                if is_scope:
                    if hasattr(node, "lineno") and hasattr(node, "end_lineno"):
                        if node.lineno <= self.target_line <= node.end_lineno:
                            scope_name = (
                                f"class {node.name}"
                                if isinstance(node, ast.ClassDef)
                                else f"def {node.name}"
                            )
                            self.current_path.append(scope_name)
                            self.best_match = list(self.current_path)
                            super().generic_visit(node)
                            self.current_path.pop()
                            return
                super().generic_visit(node)

        visitor = ScopeVisitor(line)
        visitor.visit(tree)
        if visitor.best_match:
            return ScopeRef(name=" -> ".join(visitor.best_match), kind="")
        return None

    def get_symbol_locations(
        self, file_path: FilePath, symbol: SymbolName
    ) -> ResponseDataList:
        """Stub for symbol location retrieval."""
        return ResponseDataList(values=[])

    def find_flow(
        self,
        file_path: FilePath,
        var_name: SymbolName,
        start_line: LineNumber = LineNumber(value=0),
    ) -> DataFlowList:
        if not self._fs.exists(file_path).value:
            return DataFlowList(values=[])

        source_vo: FileContentVO = self._fs.read_text(file_path)
        source = source_vo.value
        lines = source.splitlines(keepends=True)

        try:
            tree = ast.parse(source)
        except SyntaxError:
            return DataFlowList(values=[])

        vn = str(var_name)
        sl = int(start_line)
        flows = []
        target_scope: ast.AST | None = tree

        if sl > 0:

            class TargetScopeVisitor(ast.NodeVisitor):
                def __init__(self, target: LineNumber):
                    self.target = int(target)
                    self.node: ast.AST | None = None

                def generic_visit(self, node: ast.AST):
                    if isinstance(
                        node, (ast.FunctionDef, ast.AsyncFunctionDef, ast.ClassDef)
                    ):
                        lineno = getattr(node, "lineno", None)
                        end_lineno = getattr(node, "end_lineno", None)
                        if lineno is not None and end_lineno is not None:
                            if lineno <= self.target <= end_lineno:
                                self.node = node
                    super().generic_visit(node)

            tsv = TargetScopeVisitor(start_line)
            tsv.visit(tree)
            if tsv.node is not None:
                target_scope = tsv.node

        class FlowVisitor(ast.NodeVisitor):
            def visit_Name(self, node):
                if node.id == vn:
                    if hasattr(node, "lineno"):
                        line_text = lines[node.lineno - 1].strip()
                        if isinstance(node.ctx, ast.Store):
                            flows.append(
                                f"Line {node.lineno} [Assignment]: {line_text}"
                            )
                        elif isinstance(node.ctx, ast.Load):
                            flows.append(f"Line {node.lineno} [Usage]: {line_text}")
                self.generic_visit(node)

            def visit_Call(self, node):
                if isinstance(node.func, ast.Attribute) and isinstance(
                    node.func.value, ast.Name
                ):
                    if node.func.value.id == vn:
                        if hasattr(node, "lineno"):
                            line_text = lines[node.lineno - 1].strip()
                            flows.append(
                                f"Line {node.lineno} [Mutation '{node.func.attr}']: {line_text}"
                            )
                self.generic_visit(node)

        if target_scope is not None:
            FlowVisitor().visit(target_scope)

        unique_flows = list(dict.fromkeys(flows))
        unique_flows.sort(key=self._extract_lineno)
        return DataFlowList(values=unique_flows)

    def _extract_lineno(self, fstr: str) -> int:
        """Helper to extract line number from flow string."""
        try:
            return int(fstr.split("Line ")[1].split(" ")[0])
        except Exception:
            return 0

    def trace_call_chain(
        self, root_dir: DirectoryPath, target_name: SymbolName
    ) -> CallChainList:
        name = str(target_name)
        callers = []
        pattern = re.compile(rf"\b{name}\(")

        # Using the tool (infrastructure) to find files
        py_files = self._fs.glob(f"{root_dir}/**/*.py")

        for filepath in py_files:
            source_vo = self._fs.read_text(filepath)
            lines = source_vo.value.splitlines()

            for i, line in enumerate(lines):
                if pattern.search(line) and f"def {name}" not in line:
                    rel_path = self._fs.get_relative_path(
                        filepath, FilePath(value=str(root_dir))
                    )
                    callers.append(f"{rel_path}:{i + 1} -> {line.strip()}")
        return CallChainList(values=callers)

    def project_wide_rename(
        self, root_dir: DirectoryPath, old_name: SymbolName, new_name: SymbolName
    ) -> Count:
        old = str(old_name)
        new = str(new_name)
        pattern = re.compile(
            rf"""
            (
                \"\"\"(?:\\.|[^\\])*?\"\"\" |
                \'\'\'(?:\\.|[^\\])*?\'\'\' |
                \"(?:\\.|[^\"\\])*\" |
                \'(?:\\.|[^\'\\])*\' |
                \#[^\n]*
            )
            |
            \b({re.escape(old)})\b
        """,
            re.VERBOSE | re.DOTALL,
        )

        def replacer(match):
            if match.group(1) is not None:
                return match.group(1)
            return new

        py_files = self._fs.glob(f"{root_dir}/**/*.py")
        modified_count: int = 0
        for filepath in py_files:
            source_vo = self._fs.read_text(filepath)
            source = source_vo.value

            if old in source:
                new_source = pattern.sub(replacer, source)
                if new_source != source:
                    res = self._fs.write_text(filepath, FileContentVO(value=new_source))
                    if res.value:
                        modified_count += 1
        return Count(value=modified_count)

File: lint-arwaky/src-rust/capabilities/semantic_scope_analyzer.rs
// semantic_scope_analyzer — AST-based semantic scope analysis capability.
// Implements ISemanticTracerProtocol for Python code analysis.
// Uses regex-based analysis (no Python AST dependency).

use crate::taxonomy::*;
use once_cell::sync::Lazy;
use regex::Regex;
use std::collections::HashSet;

/// Regex to split names into words by camelCase, PascalCase, underscores, hyphens.
static WORD_SPLIT_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"[A-Za-z][a-z0-9]*|[A-Z]+(?=[A-Z][a-z0-9]|\b)|[0-9]+").unwrap()
});

/// Regex to detect Python function definitions.
static PY_DEF_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"^(?:async\s+)?def\s+(\w+)\s*\(").unwrap()
});

/// Regex to detect Python class definitions.
static PY_CLASS_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"^class\s+(\w+)").unwrap()
});

/// Regex for project-wide rename: matches strings, comments, and word boundaries.
static RENAME_PATTERN_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(
        r#"(?s)("""[\s\S]*?"""|'''[\s\S]*?'''|"(?:\\.|[^"\\])*"|'(?:\\.|[^'\\])*'|#[^\n]*)|\b(\w+)\b"#,
    )
    .unwrap()
});

/// AST-based semantic scope analyzer for Python code.
pub struct SemanticScopeAnalyzer;

impl SemanticScopeAnalyzer {
    pub fn new() -> Self {
        Self
    }

    /// Split a name into lowercase words by camelCase, PascalCase, underscores, hyphens.
    fn split_words(name: &str) -> Vec<String> {
        WORD_SPLIT_RE
            .find_iter(name)
            .map(|m| m.as_str().to_lowercase())
            .collect()
    }

    /// Capitalize the first character of a string.
    fn capitalize(s: &str) -> String {
        let mut c = s.chars();
        match c.next() {
            None => String::new(),
            Some(f) => f.to_uppercase().collect::<String>() + c.as_str(),
        }
    }

    /// Generate snake_case, camelCase, PascalCase, SCREAMING_SNAKE variants.
    pub fn get_variant_dict(&self, name: &SymbolName) -> serde_json::Value {
        let n = &name.value;
        let words = Self::split_words(n);

        if words.is_empty() {
            return serde_json::json!({
                "snake_case": n,
                "pascal_case": n,
                "camel_case": n,
                "screaming_snake": n.to_uppercase()
            });
        }

        let snake_case = words.join("_");
        let pascal_case: String = words.iter().map(|w| Self::capitalize(w)).collect();
        let camel_case = if words.len() > 1 {
            let mut c = words[0].clone();
            for w in &words[1..] {
                c.push_str(&Self::capitalize(w));
            }
            c
        } else {
            words[0].clone()
        };
        let screaming_snake = snake_case.to_uppercase();

        serde_json::json!({
            "snake_case": snake_case,
            "camel_case": camel_case,
            "pascal_case": pascal_case,
            "screaming_snake": screaming_snake
        })
    }

    /// Returns a SymbolNameList of all possible naming variants.
    pub fn build_variants(&self, name: &SymbolName) -> SymbolNameList {
        let dict = self.get_variant_dict(name);
        let mut results: HashSet<String> = HashSet::new();
        results.insert(name.value.clone());

        if let Some(v) = dict.get("snake_case").and_then(|v| v.as_str()) {
            results.insert(v.to_string());
        }
        if let Some(v) = dict.get("camel_case").and_then(|v| v.as_str()) {
            results.insert(v.to_string());
        }
        if let Some(v) = dict.get("pascal_case").and_then(|v| v.as_str()) {
            results.insert(v.to_string());
        }
        if let Some(v) = dict.get("screaming_snake").and_then(|v| v.as_str()) {
            results.insert(v.to_string());
        }
        // Add kebab-case variant
        if let Some(v) = dict.get("snake_case").and_then(|v| v.as_str()) {
            results.insert(v.replace('_', "-"));
        }

        SymbolNameList {
            values: results.into_iter().map(SymbolName::new).collect(),
        }
    }

    /// AST-based enclosing scope lookup.
    pub fn get_enclosing_scope(&self, file_path: &FilePath, line: LineNumber) -> Option<ScopeRef> {
        let target_line = line.value;
        if target_line <= 0 {
            return None;
        }

        let content = std::fs::read_to_string(file_path.to_string()).ok()?;
        let lines: Vec<&str> = content.lines().collect();

        // Stack of (scope_name, start_line, end_line)
        let mut scope_stack: Vec<(String, i64, i64)> = Vec::new();
        let mut best_match: Vec<String> = Vec::new();

        for (i, raw_line) in lines.iter().enumerate() {
            let current_line = (i as i64) + 1;
            let stripped = raw_line.trim();

            // Pop expired scopes
            scope_stack.retain(|&(_, _, end)| end >= current_line);

            // Detect new scope
            if let Some(cap) = PY_CLASS_RE.captures(stripped) {
                if let Some(name) = cap.get(1) {
                    let scope_name = format!("class {}", name.as_str());
                    let indent = raw_line.len() - raw_line.trim_start().len();
                    let mut end_line = lines.len() as i64;
                    for j in (i + 1)..lines.len() {
                        let l = lines[j];
                        if !l.trim().is_empty() {
                            let l_indent = l.len() - l.trim_start().len();
                            if l_indent <= indent
                                && (PY_CLASS_RE.is_match(l.trim()) || PY_DEF_RE.is_match(l.trim()))
                            {
                                end_line = j as i64;
                                break;
                            }
                        }
                    }
                    scope_stack.push((scope_name, current_line, end_line));
                    if current_line <= target_line && target_line <= end_line {
                        best_match = scope_stack.iter().map(|(n, _, _)| n.clone()).collect();
                    }
                }
            } else if let Some(cap) = PY_DEF_RE.captures(stripped) {
                if let Some(name) = cap.get(1) {
                    let scope_name = format!("def {}", name.as_str());
                    let indent = raw_line.len() - raw_line.trim_start().len();
                    let mut end_line = lines.len() as i64;
                    for j in (i + 1)..lines.len() {
                        let l = lines[j];
                        if !l.trim().is_empty() {
                            let l_indent = l.len() - l.trim_start().len();
                            if l_indent <= indent
                                && (PY_CLASS_RE.is_match(l.trim()) || PY_DEF_RE.is_match(l.trim()))
                            {
                                end_line = j as i64;
                                break;
                            }
                        }
                    }
                    scope_stack.push((scope_name, current_line, end_line));
                    if current_line <= target_line && target_line <= end_line {
                        best_match = scope_stack.iter().map(|(n, _, _)| n.clone()).collect();
                    }
                }
            }
        }

        if !best_match.is_empty() {
            Some(ScopeRef {
                name: best_match.join(" -> "),
                kind: String::new(),
                file: Some(file_path.clone()),
                start_line: None,
                end_line: None,
            })
        } else {
            None
        }
    }

    /// Stub for symbol location retrieval.
    pub fn get_symbol_locations(
        &self,
        _file_path: &FilePath,
        _symbol: &SymbolName,
    ) -> Vec<serde_json::Value> {
        Vec::new()
    }

    /// AST-based data flow analysis (assignments, usages, mutations).
    pub fn find_flow(
        &self,
        file_path: &FilePath,
        var_name: &SymbolName,
        start_line: LineNumber,
    ) -> DataFlowList {
        let content = match std::fs::read_to_string(file_path.to_string()) {
            Ok(c) => c,
            Err(_) => return DataFlowList { values: Vec::new() },
        };

        let lines: Vec<&str> = content.lines().collect();
        let vn = &var_name.value;
        let sl = start_line.value as usize;

        // Determine target scope bounds
        let (scope_start, scope_end) = if sl > 0 {
            let mut start = sl;
            let mut end = lines.len();

            for (i, raw_line) in lines.iter().enumerate() {
                let line_no = i + 1;
                let stripped = raw_line.trim();
                if line_no <= sl {
                    if PY_DEF_RE.is_match(stripped) || PY_CLASS_RE.is_match(stripped) {
                        start = line_no;
                    }
                } else if !stripped.is_empty() {
                    let indent = raw_line.len() - raw_line.trim_start().len();
                    if indent == 0 && (PY_DEF_RE.is_match(stripped) || PY_CLASS_RE.is_match(stripped)) {
                        end = line_no;
                        break;
                    }
                }
            }

            (Some(start), Some(end))
        } else {
            (None, None)
        };

        let assign_re = Regex::new(&format!(r"\b{}\s*=", regex::escape(vn))).unwrap();
        let mutation_re = Regex::new(&format!(r"\b{}\.\w+", regex::escape(vn))).unwrap();
        let word_re = Regex::new(&format!(r"\b{}\b", regex::escape(vn))).unwrap();

        let mut flows: Vec<ErrorMessage> = Vec::new();
        let mut seen: HashSet<String> = HashSet::new();

        for (i, line_str) in lines.iter().enumerate() {
            let line_no = (i as i64) + 1;

            if let Some(s) = scope_start {
                if (line_no as usize) < s {
                    continue;
                }
            }
            if let Some(e) = scope_end {
                if (line_no as usize) > e {
                    break;
                }
            }

            if !word_re.is_match(line_str) {
                continue;
            }

            let stripped = line_str.trim();
            let entry_str = if mutation_re.is_match(line_str) {
                let method = mutation_re
                    .find(line_str)
                    .and_then(|m| m.as_str().split('.').nth(1))
                    .unwrap_or("mutation");
                format!("Line {} [Mutation '{}']: {}", line_no, method, stripped)
            } else if assign_re.is_match(line_str) {
                format!("Line {} [Assignment]: {}", line_no, stripped)
            } else {
                format!("Line {} [Usage]: {}", line_no, stripped)
            };

            if seen.contains(&entry_str) {
                continue;
            }
            seen.insert(entry_str.clone());
            flows.push(ErrorMessage::new(entry_str));
        }

        flows.sort_by(|a, b| extract_lineno(&a.value).cmp(&extract_lineno(&b.value)));
        DataFlowList { values: flows }
    }

    /// Project-wide call chain tracing.
    pub fn trace_call_chain(
        &self,
        root_dir: &DirectoryPath,
        target_name: &SymbolName,
    ) -> Vec<SymbolName> {
        let name = &target_name.value;
        let call_re = match Regex::new(&format!(r"\b{}\s*\(", regex::escape(name))) {
            Ok(r) => r,
            Err(_) => return Vec::new(),
        };
        let def_re = match Regex::new(&format!(r"def\s+{}\s*\(", regex::escape(name))) {
            Ok(r) => r,
            Err(_) => return Vec::new(),
        };

        let mut callers: Vec<SymbolName> = Vec::new();

        if let Ok(py_files) = collect_py_files(&root_dir.value) {
            for filepath in &py_files {
                let content = match std::fs::read_to_string(filepath) {
                    Ok(c) => c,
                    Err(_) => continue,
                };

                for (i, line) in content.lines().enumerate() {
                    if call_re.is_match(line) && !def_re.is_match(line) {
                        let rel_path = filepath
                            .strip_prefix(&root_dir.value)
                            .unwrap_or(filepath);
                        callers.push(SymbolName::new(format!(
                            "{}:{} -> {}",
                            rel_path,
                            i + 1,
                            line.trim()
                        )));
                    }
                }
            }
        }

        callers
    }

    /// Project-wide symbol rename with string/comment awareness.
    pub fn project_wide_rename(
        &self,
        root_dir: &DirectoryPath,
        old_name: &SymbolName,
        new_name: &SymbolName,
    ) -> Count {
        let old = &old_name.value;
        let new = &new_name.value;

        let pattern = match Regex::new(&format!(
            r#"(?s)("""[\s\S]*?"""|'''[\s\S]*?'''|"(?:\\.|[^"\\])*"|'(?:\\.|[^'\\])*'|#[^\n]*)|\b{}\b"#,
            regex::escape(old)
        )) {
            Ok(r) => r,
            Err(_) => return Count::new(0),
        };

        let mut modified_count: i64 = 0;

        if let Ok(py_files) = collect_py_files(&root_dir.value) {
            for filepath in &py_files {
                if let Ok(content) = std::fs::read_to_string(filepath) {
                    if content.contains(old) {
                        let new_source = pattern
                            .replace_all(&content, |caps: &regex::Captures| {
                                if caps.get(1).is_some() {
                                    caps.get(0).unwrap().as_str().to_string()
                                } else {
                                    new.to_string()
                                }
                            })
                            .to_string();
                        if new_source != content {
                            if std::fs::write(filepath, &new_source).is_ok() {
                                modified_count += 1;
                            }
                        }
                    }
                }
            }
        }

        Count::new(modified_count)
    }
}

/// Extract line number from a flow string like "Line 5 [Assignment]: ..."
fn extract_lineno(fstr: &str) -> i64 {
    fstr.split("Line ")
        .nth(1)
        .and_then(|s| s.split(' ').next())
        .and_then(|s| s.parse().ok())
        .unwrap_or(0)
}

/// Recursively collect all .py files under a directory.
fn collect_py_files(dir: &str) -> Result<Vec<String>, std::io::Error> {
    let mut files = Vec::new();
    collect_py_files_recursive(dir, &mut files)?;
    Ok(files)
}

fn collect_py_files_recursive(dir: &str, files: &mut Vec<String>) -> Result<(), std::io::Error> {
    for entry in std::fs::read_dir(dir)? {
        let entry = entry?;
        let path = entry.path();
        if path.is_dir() {
            collect_py_files_recursive(&path.to_string_lossy(), files)?;
        } else if path.extension().map(|e| e == "py").unwrap_or(false) {
            files.push(path.to_string_lossy().to_string());
        }
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_split_words() {
        let words = SemanticScopeAnalyzer::split_words("getVariantDict");
        assert_eq!(words, vec!["get", "variant", "dict"]);
    }

    #[test]
    fn test_split_words_underscore() {
        let words = SemanticScopeAnalyzer::split_words("snake_case_name");
        assert_eq!(words, vec!["snake", "case", "name"]);
    }

    #[test]
    fn test_extract_lineno() {
        assert_eq!(extract_lineno("Line 5 [Assignment]: x = 1"), 5);
        assert_eq!(extract_lineno("no line here"), 0);
    }

    #[test]
    fn test_get_variant_dict() {
        let analyzer = SemanticScopeAnalyzer::new();
        let name = SymbolName::new("getUserName");
        let dict = analyzer.get_variant_dict(&name);
        assert_eq!(dict["snake_case"], "get_user_name");
        assert_eq!(dict["camel_case"], "getUserName");
        assert_eq!(dict["pascal_case"], "GetUserName");
        assert_eq!(dict["screaming_snake"], "GET_USER_NAME");
    }
}
